**NOTES**
- Topic: How advances in contact tracing technology used during the COVID-19 pandemic can be used to predict future disease epidemics and pandemics
- Imported datasets are from your Google Drive

**Prepared by:**
- Email: wiener baalls

# 0. Import packages

In [1]:
# @title Import needed libraries
import pandas as pd
import numpy as np

# @title Import Google Drive directory
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# 1. Import and preview the DataFrame

In [2]:
# @title Import DataFrame
file_path = '/content/drive/MyDrive/Colab Notebooks/colab-files/figshare_taiwan_covid.xlsx'
df_ind = pd.read_excel(file_path, sheet_name='Individual Data')

In [5]:
# @title View a preview of the DataFrame
df_ind.tail(5)

id abroad/local infection_source visited_countries/cities  \
573   5       Abroad            China                    Wuhan   
574   4       Abroad            China            Wuhan, Europe   
575   3       Abroad            China                    Wuhan   
576   2       Abroad            China         Travel to Taiwan   
577   1       Abroad            China                    Wuhan   

    date_of_departure_1 date_of_arrival_1 visited_country/city_1  \
573                 NaN               NaN                  Wuhan   
574          2020-01-13               NaN                  Wuhan   
575                 NaN               NaN                  Wuhan   
576                 NaN               NaN       Travel to Taiwan   
577                 NaN               NaN                  Wuhan   

    date_of_departure_2 date_of_arrival_2 visited_country/city_2  ...  \
573                 NaN               NaN                    NaN  ...   
574          2020-01-15        2020-01-16                 Europe  ...   
575                 NaN               NaN                    NaN  ...   
576                 NaN               NaN                    NaN  ...   
577                 NaN               NaN                    NaN  ...   

    date_of_contact_with_infected_case source_infected_cases_all  \
573                                NaN                       NaN   
574                                NaN                       NaN   
575                                NaN                       NaN   
576                                NaN                       NaN   
577                                NaN                       NaN   

    source_infected_case earliest_infection_date negative_test_date_1  \
573                  NaN                     NaT                  NaT   
574                  NaN                     NaT                  NaT   
575                  NaN                     NaT                  NaT   
576                  NaN                     NaT                  NaT   
577                  NaN                     NaT                  NaT   

    negative_test_date_2 negative_test_date_3 negative_test_date_4  \
573                  NaT                  NaT                  NaT   
574                  NaT                  NaT                  NaT   
575                  NaT                  NaT                  NaT   
576                  NaT                  NaT                  NaT   
577                  NaT                  NaT                  NaT   

                                                detail  \
573  business;\nOn the 26th, there were 119 newly n...   
574  travelling;\n新增1例境外移入嚴重特殊傳染性肺炎病例，為北部50多歲女性，1月1...   
575  business; \nAs of 4:00 pm on January 24, there...   
576  travelling\n50多歲男性國人(案2)，皆為1月21日入境；案1為1月23日因發燒...   
577  business\n50多歲中國籍女性(案1)，皆為1月21日入境；案2於1月20日出現感冒...   

                                                source  
573  https://www.cdc.gov.tw/Bulletin/Detail/d-4jLlO...  
574  https://www.cdc.gov.tw/Bulletin/Detail/d-4jLlO...  
575  https://www.cdc.gov.tw/Bulletin/Detail/d-4jLlO...  
576  https://www.cdc.gov.tw/Bulletin/Detail/d-4jLlO...  
577  https://www.cdc.gov.tw/Bulletin/Detail/d-4jLlO...  

[5 rows x 40 columns]

In [6]:
df_ind.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 578 entries, 0 to 577
Data columns (total 40 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   id                                  578 non-null    int64         
 1   abroad/local                        577 non-null    object        
 2   infection_source                    474 non-null    object        
 3   visited_countries/cities            472 non-null    object        
 4   date_of_departure_1                 440 non-null    object        
 5   date_of_arrival_1                   39 non-null     object        
 6   visited_country/city_1              471 non-null    object        
 7   date_of_departure_2                 77 non-null     object        
 8   date_of_arrival_2                   9 non-null      object        
 9   visited_country/city_2              30 non-null     object        
 10  date_of_departure_3       

# 2. Let him cook

## 1 Measures of Central Tendency

In [ ]:
# @title Create sample DataFrames

df1 = pd.DataFrame({"X": [54, 54, 54, 55, 56, 57, 57, 58, 58, 60, 60]})
df2 = pd.DataFrame({"X": [1, 2, 54, 54, 54, 55, 56, 57, 57, 58, 58, 60, 60, 200, 250]})

In [ ]:
# @title Mean, Median, Mode

mean1 = df1["X"].mean().round(1)
median1 = df1["X"].median()
mode1 = df1["X"].mode()[0]

print(f"[DF1] mean = {mean1}, median = {median1}, mode = {mode1}")

mean2 = df2["X"].mean().round(1)
median2 = df2["X"].median()
mode2 = df2["X"].mode()[0]

print(f"[DF2] mean = {mean2}, median = {median2}, mode = {mode2}")

[DF1] mean = 56.6, median = 57.0, mode = 54
[DF2] mean = 71.7, median = 57.0, mode = 54


In [ ]:
# @title Trimmed Mean

from scipy.stats import trim_mean

trimmed_mean2 = trim_mean(df2["X"], proportiontocut=0.15).round(1)

print(f"[DF2] trimmed mean (removed 15%) = {trimmed_mean2}")

[DF2] trimmed mean (removed 15%) = 56.6


In [4]:
?trim_mean


# penois

Object `trim_mean` not found.


## 2 Measures of Dispersion

In [ ]:
# @title Create sample DataFrames

df1 = pd.DataFrame({"X": [54, 54, 54, 55, 56, 57, 57, 58, 58, 60, 60]})
df2 = pd.DataFrame({"X": [1, 2, 54, 54, 54, 55, 56, 57, 57, 58, 58, 60, 60, 200, 250]})

In [ ]:
# @title Range

range1 = df1["X"].max() - df1["X"].min()
print(f"[DF1] range = {range1}")

range2 = df2["X"].max() - df2["X"].min()
print(f"[DF2] range = {range2}")

[DF1] range = 6
[DF2] range = 249


In [ ]:
# @title Interquartile Range

q1_1 = df1["X"].quantile(0.25)
q3_1 = df1["X"].quantile(0.75)
iqr_1 = q3_1 - q1_1
print(f"[DF1] IQR = {q3_1} - {q1_1} = {iqr_1}")

q1_2 = df2["X"].quantile(0.25)
q3_2 = df2["X"].quantile(0.75)
iqr_2 = q3_2 - q1_2
print(f"[DF2] IQR = {q3_2} - {q1_2} = {iqr_2}")

[DF1] IQR = 58.0 - 54.5 = 3.5
[DF2] IQR = 59.0 - 54.0 = 5.0


The IQR in the notebook differs from the ones in the slide due to how pandas computes for the quantile. See the `interpolation` parameter.

In [ ]:
?df1.quantile

In [ ]:
# @title Variance and Standard Deviation

mean1 = df1["X"].mean()
variance1 = df1["X"].var()
std1 = df1["X"].std()
print(f"[DF1] mean = {mean1:.2f}, variance = {variance1:.2f}, standard deviation = {std1:.2f}")

mean2 = df2["X"].mean()
variance2 = df2["X"].var()
std2 = df2["X"].std()
print(f"[DF2] mean = {mean2:.2f}, variance = {variance2:.2f}, standard deviation = {std2:.2f}")

[DF1] mean = 56.64, variance = 5.05, standard deviation = 2.25
[DF2] mean = 71.73, variance = 4332.50, standard deviation = 65.82


DF1 Interpretation
- The data points in DF1 are tightly clustered around the mean.
- The standard deviation of 2.25 means most values fall roughly within ±2.25 units of the mean (i.e., between about 54.4 and 58.9).
- Variance is low, which implies that DF1's data is more reliable/stable.

DF2 Interpretation
- DF2's data points are much more spread out around the mean.
- A standard deviation of 65.82 means values vary widely, typically within ±65.82 units of the mean (so roughly between 6 and 138).
- The very high variance confirms there is huge variability or inconsistency in DF2's data.
- DF2's data shows greater diversity or possibly the presence of outliers/extreme values

## 3 Correlation

In [ ]:
# @title Create sample DataFrames

data = {
    "height": [154, 160, 167, 152, 175, 161, 168],
    "weight": [60, 70, 65, 55, 85, 80, 75]
}

df = pd.DataFrame(data)
df

height  weight
0     154      60
1     160      70
2     167      65
3     152      55
4     175      85
5     161      80
6     168      75

### 3.1 Pearson Correlation

$$r_{xy} = \frac{cov(x,y)}{\sigma_x \cdot  \sigma_y} = \frac{\sum_{i=1}^{n}(x_i-\overline{x})(y_i-\overline{y})}{\sqrt{\sum_{i=1}^{n}(x_i-\overline{x})^2} \cdot \sqrt{\sum_{i=1}^{n}(y_i-\overline{y})^2}}$$

In [ ]:
# @title Manual Computaion: Expand DataFrame

df["x_minus_mean"] = df["height"] - df["height"].mean()
df["y_minus_mean"] = df["weight"] - df["weight"].mean()
df["product"] = df["x_minus_mean"] * df["y_minus_mean"]
df["x_minus_mean_squared"] = df["x_minus_mean"] ** 2
df["y_minus_mean_squared"] = df["y_minus_mean"] ** 2
df

height  weight  x_minus_mean  y_minus_mean     product  \
0     154      60     -8.428571         -10.0   84.285714   
1     160      70     -2.428571           0.0   -0.000000   
2     167      65      4.571429          -5.0  -22.857143   
3     152      55    -10.428571         -15.0  156.428571   
4     175      85     12.571429          15.0  188.571429   
5     161      80     -1.428571          10.0  -14.285714   
6     168      75      5.571429           5.0   27.857143   

   x_minus_mean_squared  y_minus_mean_squared  
0             71.040816                 100.0  
1              5.897959                   0.0  
2             20.897959                  25.0  
3            108.755102                 225.0  
4            158.040816                 225.0  
5              2.040816                 100.0  
6             31.040816                  25.0

In [ ]:
# @title Manual Computation: Compute for Pearson Correlation

numerator = df["product"].sum()
denominator = np.sqrt(df["x_minus_mean_squared"].sum()) * np.sqrt(df["y_minus_mean_squared"].sum())
pearson = numerator / denominator

print(f"r = {pearson:.3f}")

r = 0.796


In [ ]:
# @title Automatic Computation: Using pandas `corr()` function

pearson = df["height"].corr(df["weight"], method="pearson")
print(f"r = {pearson:.3f}")

r = 0.796


### 3.2 Spearman Correlation

$$r_{xy} = \frac{cov(rank(x), rank(y))}{\sigma_{rank(x)} \cdot  \sigma_{rank(y)}} = 1 - \frac{6\sum{d_i^2}}{n(n^2-1)}$$

In [ ]:
# @title Manual Computaion: Expand DataFrame

df["rank_x"] = df["height"].rank()
df["rank_y"] = df["weight"].rank()
df["d"] = df["rank_x"] - df["rank_y"]
df["d_squared"] = df["d"] ** 2

df

height  weight  rank_x  rank_y    d  d_squared
0     154      60     2.0     2.0  0.0        0.0
1     160      70     3.0     4.0 -1.0        1.0
2     167      65     5.0     3.0  2.0        4.0
3     152      55     1.0     1.0  0.0        0.0
4     175      85     7.0     7.0  0.0        0.0
5     161      80     4.0     6.0 -2.0        4.0
6     168      75     6.0     5.0  1.0        1.0

In [ ]:
# @title Manual Computation: Compute for Spearman Correlation

numerator = 6 * df["d_squared"].sum()
denominator = len(df) * (len(df) ** 2 - 1)
spearman = 1 - (numerator / denominator)

print(f"r = {spearman:.3f}")

r = 0.821


In [ ]:
# @title Automatic Computation: Using pandas `corr()` function

spearman = df["height"].corr(df["weight"], method="spearman")
print(f"r = {spearman:.3f}")

r = 0.821
